cmd/device-worker/main.go

```diff
@@ -147,9 +152,46 @@ func main() {
 
 	setupSignalHandler(metricsStore)
 
+	go listenStartGracefulRebootChannel(wl, dataMonitor, systemMetricsWatcher, metricsStore, hbs,
+		gracefulRebootChannel, deviceOs)
+
 	if err := s.Serve(l); err != nil {
 		log.Fatalf("cannot start worker server, err: %v", err)
 	}
+
+}
+
+func listenStartGracefulRebootChannel(wl *workload2.WorkloadManager, dataMonitor *datatransfer.Monitor,
+	systemMetricsWatcher *metrics.SystemMetrics, metricsStore *metrics.TSDB, hbs *heartbeat2.Heartbeat,
+	gracefulRebootChannel chan struct{}, deviceOs *os2.OS) {
+	// listen to the channel for getting StartGracefulReboot signal
+	for {
+		select {
+		case <-gracefulRebootChannel:
+			log.Info("A graceful reboot request was received")
+			if err := wl.StopWorkloads(); err != nil {
+				log.Fatalf("cannot graceful reboot the workloads: %v", err)
+			}
+
+			if err := dataMonitor.Deregister(); err != nil {
+				log.Fatalf("cannot graceful reboot dataMonitor: %v", err)
+			}
+
+			if err := systemMetricsWatcher.Deregister(); err != nil {
+				log.Fatalf("cannot graceful reboot systemMetricsWatcher: %v", err)
+			}
+
+			if err := metricsStore.Deregister(); err != nil {
+				log.Fatalf("cannot graceful reboot metricsStore: %v", err)
+			}
+
+			if err := hbs.Deregister(); err != nil {
+				log.Fatalf("cannot graceful reboot the heartbeat service: %v", err)
+			}
+
+			gracefulRebootChannel <- struct{}{}
```

Comment:
can't this lead to another invocation of this for loop? 
shouldn't we break outside or use a dedicated channel for graceful shutdown completion?

Reply:
I think Moti is right that there might be some unexpected behavior as a result of that call. I suspect there might be a race condition and in some cases that may result in an infinite loop.

Reply:
I tested it in a playground env 
but for making it clear I will add